Fixed #36410 -- Added named template partials to DTL

FarhanAliRaza opened the pull request:
- Updated `Engine.get_template` to load a partial template
- Added `partialdef `and `partial` template tags
- Added documentation for the partials tags
- Added tests for the partial tags

#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-36410

#### Branch description
Provide a concise overview of the issue or rationale behind the proposed changes.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

nessita commented on django/template/backends/django.py:
```diff
@@ -75,10 +75,26 @@ def from_string(self, template_code):
+        partial_contents = extra_data.get("template-partials", {})
+
+        try:
+            partial = partial_contents[partial_name]
```
Is there a change that `partial_contents` is None or any other non-mapping type? If yes, we need a test for those cases and also to catch `TypeError` (since `None["foo"]` raises `TypeError` and not `KeyError`).

nessita commented on django/template/backends/django.py:
```diff
@@ -75,10 +75,26 @@ def from_string(self, template_code):
+
+        try:
+            partial = partial_contents[partial_name]
+        except KeyError:
+            # Partial not found on this template.
+            raise TemplateDoesNotExist(partial_name, tried=[template_name])
```
I don't think this block is covered by a test, I removed it and no test fails.

nessita commented on django/template/backends/django.py:
```diff
@@ -75,10 +75,26 @@ def from_string(self, template_code):
+        except KeyError:
+            # Partial not found on this template.
+            raise TemplateDoesNotExist(partial_name, tried=[template_name])
+        partial.engine = self.engine
```
Same here, I don't think this line is covered by a test, I removed it and no test fails.

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+
+    def render(self, context):
+        """Set content into context and return empty string"""
+        if self.inline:
+            return self.nodelist.render(context)
+        else:
+            return ""
```
```suggestion
         return self.nodelist.render(context) if self.inline else ""
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    # Parse the tag
```
```suggestion
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+    # Parse the tag
+    tokens = token.split_contents()
+
+    # check we have the expected number of tokens before trying to assign them
+    # via indexes
```
```suggestion
    # Check the number of tokens before trying to assign them via indexes.
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+
+    # check we have the expected number of tokens before trying to assign them
+    # via indexes
+    if len(tokens) not in (2, 3):
+        raise TemplateSyntaxError(
+            "%r tag requires 2-3 arguments" % token.contents.split()[0]
+        )
+
+    partial_name = tokens[1]
+
+    try:
+        inline = tokens[2]
+    except IndexError:
+        # the inline argument is optional, so fallback to not using it
+        inline = False
```
```suggestion
    if (tokens_len := len(tokens)) not in (2, 3):
        name = token.contents.split()[0]
        raise TemplateSyntaxError(f"{name} tag requires 2-3 arguments")

    partial_name = tokens[1]
    if tokens_len > 2:
        inline = tokens[2]
    else:
        inline = False
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+        inline = False
+
+    if inline and inline != "inline":
+        warnings.warn(
+            "The 'inline' argument does not have any parameters; "
+            "either use 'inline' or remove it completely.",
+            DeprecationWarning,
+        )
```
What does it mean that this raises a deprecation warning? When is the deprecation ending? Why not raise an error here since this is a new feature?

Also we can move this check inside the `if` block right after the `inline = tokens[2]` sentence.

nessita replied:
Adding to the above, I'm not sure this is tested at all :thinking:

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+        return self.nodelist.render(context)
+
+    def render(self, context):
+        "Display stage -- can be called many times"
```
This is not following the documented style for docstrings, it should use triple double quotes, and ideally it would only be defined if the docstring adds valuable information not clear from the implementation.

FarhanAliRaza replied:
I removed it.  As it is not needed.

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+            sspos, sepos = m.span()
+            starter, name, inline = m.groups()
+            endm = partial_end_tag_re.search(full_source, sepos + 1)
+            assert endm, "End tag must be present"
```
While I understand how using raw asserts to verify "truths" is handy, the Django code does not use this pattern. Could you please remove it to keep a consistent style with the current code base?

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+            assert endm, "End tag must be present"
+            espos, eepos = endm.span()
+            if name == partial_name:
+                # Include the full partial definition from opening to closing tag
```
Comments that are sentences should end with a period.
```suggestion
                # Include the full partial definition from opening to closing tag.
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
```
Is there a change that this raises `KeyError` somehow?

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return template.get_exception_info(exception, token)
+
+    def find_partial_source(self, full_source, partial_name):
```
This method may not be fully tested, could you please point me to the tests covering the non-trivial logic described here?

In particular I think we need a new test file covering all the corner cases in `syntax_tests/test_partials` or similar.

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+            DeprecationWarning,
+        )
+
+    # Parse the content until the end tag
```
```suggestion
    # Parse the content until the end tag.
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+    if endpartial.contents not in acceptable_endpartials:
+        parser.invalid_block_tag(endpartial, "endpartialdef", acceptable_endpartials)
+
+    # Store the partial nodelist in the parser.extra_data attribute,
```
```suggestion
    # Store the partial nodelist in the parser.extra_data attribute.
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+    return DefinePartialNode(partial_name, inline, nodelist)
+
+
+class SubDictionaryWrapper:
```
Would this be better placed in the django.template.utils module? This could also use a test case covering the functionality?

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+
+        {% partial partial_name %}
+    """
+    # Parse the tag
```
This is also not adding much info to the code.
```suggestion
```

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+            )
+
+
+# Define the partial tag to render the partial content.
```
We could remove this comment.
```suggestion
```

nessita commented on django/test/utils.py:
```diff
@@ -25,6 +25,7 @@
 from django.db import DEFAULT_DB_ALIAS, connections, reset_queries
 from django.db.models.options import Options
 from django.template import Template
+from django.template.defaulttags import TemplateProxy
```
Shall we move the `TemplateProxy` to this file instead? :thought_balloon: 

I see there could be a circular dependency between defaulttags and utils, which feels like a code smell. Can you think of alternatives to clean defaulttags a bit and leave the code the more focused for the new tags and have helpers somewhere else?

carltongibson replied:
@nessita I'm not particularly fussed about the name. If you prefer `TemplatePartial` then ok, maybe. 🤷 (The original name was because it proxies for a template, providing the ability to bind the context to the node list when rendering) 

I'm sceptical about using inheritance here. Bar render then implementations aren't shared. (And render is very few lines.) We not specialising. Etc. (I.e. the positive reasons for subclassing aren't in play. All there is, is maybe save a few lines of typing, which isn't a good reason.)

nessita commented on docs/ref/templates/api.txt:
```diff
@@ -158,6 +158,29 @@ overridden by what's passed by
+
+    .. code-block:: python
+
+        # Load entire template
```
```suggestion
        # Load an entire template.
```

nessita commented on docs/ref/templates/api.txt:
```diff
@@ -158,6 +158,29 @@ overridden by what's passed by
+        # Load entire template
+        template = engine.get_template("template.html")
+
+        # Load specific partial from template
```
```suggestion
        # Load a specific fragment from a template.
```

nessita commented on docs/ref/templates/api.txt:
```diff
@@ -158,6 +158,29 @@ overridden by what's passed by
+        # Load specific partial from template
+        partial = engine.get_template("template.html#partial_name")
+
+    See :ref:`template-partials` for more information about defining and using
+    template partials.
+
+    When loading partials, the method returns a template object that behaves 
+    like a regular :class:`Template` but contains only the partial's content.
```
I would swap these two paragraphs:
```suggestion
    When loading partials, this method returns a template object that behaves 
    like a regular :class:`Template` but contains only the partial's content.

    See :ref:`template-partials` for more information about defining and using
    template partials.
```

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -957,6 +957,83 @@ output (as a string) inside a variable. This is useful if you want to use
+        <button class="btn">{{ button_text }}</button>
+    {% endpartialdef %}
+
+    {% partial button %}
+    {% partial button %}
```
Why two? shouldn't `button` be a string to not make the docs confusing? How about `button_text` , where is that defined?

Side note: I know this is a ref doc so we shouldn't go into too much detail about examples, but we either provide a simple but functional example or we link to a how-to or a more hands on doc (or both!).

FarhanAliRaza replied:
<img width="786" height="288" alt="image" src="https://github.com/user-attachments/assets/24e52196-3bec-4ef3-8251-c0a5785a73e1" />
Something like this.

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -957,6 +957,83 @@ output (as a string) inside a variable. This is useful if you want to use
+
+.. versionadded:: 6.0
+
+Defines a reusable template partial that can be rendered multiple times within
```
The word "multiple" here may imply that a user **** to use the partial multiple times. Could we soften this up to be more inviting and less strict?
```suggestion
Defines a reusable template fragment that can be rendered multiple times within
```

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -957,6 +957,83 @@ output (as a string) inside a variable. This is useful if you want to use
+.. versionadded:: 6.0
+
+Defines a reusable template partial that can be rendered multiple times within
+the same template or accessed directly via template loading.
```
```suggestion
the same template or accessed directly via template loading or inclusion.
```

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -957,6 +957,83 @@ output (as a string) inside a variable. This is useful if you want to use
+        <!-- partial content -->
+    {% endpartialdef %}
+
+The ``{% partialdef %}`` tag can be used with one or two arguments.
+The arguments are:
```
Always try to use as much as possible of the 79cols + newline.
```suggestion
The ``{% partialdef %}`` tag can be used with one or two arguments. The
arguments are:
```

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -957,6 +957,83 @@ output (as a string) inside a variable. This is useful if you want to use
+The ``{% partialdef %}`` tag can be used with one or two arguments.
+The arguments are:
+
+================  =============================================================
+Argument          Description
+================  =============================================================
+``partial_name``  String identifier for the partial (required).
+``inline``        The word ``inline``, which if given, renders the partial
+                  content immediately in addition to defining it.
+================  =============================================================
```
Could we use something else other than a table? We should try to mimic what other tags do, see for example what `include` does with `only`.

nessita commented on tests/template_tests/test_partials.py:
```diff
@@ -0,0 +1,303 @@
+        {% partial testing-partial %}
+        """
+
+        engine = engines["django"]
+        t = engine.from_string(template)
```
Given the repetition of this two lines, I think the test case could benefit from a property like:

```python
@property
def engine(self):
    return engines["django"]
```
And then, in every test:
```suggestion
        t = self.engine.from_string(template)
```

nessita commented on tests/template_tests/test_partials.py:
```diff
@@ -0,0 +1,303 @@
+        urls_module = ModuleType("partial_test_urls")
+        urls_module.urlpatterns = [path("sample/", sample_view, name="sample-view")]
+
+        with override_settings(
+            ROOT_URLCONF=urls_module,
+        ):
```
This wrapping is not needed, could you please double check the code to ensure lines shorter that 89cols fit in one line?
```suggestion
        with override_settings(ROOT_URLCONF=urls_module):
```

carltongibson commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+        self.nodelist = nodelist
+
+    def render(self, context):
+        """Set content into context and return empty string"""
```
This docstring isn't right. (Historically it was, but then we added the `inline` option, and it should have been corrected then. 

I'd might remove it at this point.

carltongibson commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+    return DefinePartialNode(partial_name, inline, nodelist)
+
+
+class SubDictionaryWrapper:
```
@nessita There were tests for this, accessing a partial before it's defined, which show how it's needed, but there's _integration tests_ if you will. 

@ngnpope I don't know what your bandwidth is like, but I thought of you as we were playing with the need to defer the dictionary access with this class. It's the sort of thing where I might expect you to have good input, if you did fancy a glance. Is there some much nicer way of getting the same result that we just didn't at? (I think it's fine™ if you don't have capacity, but I thought it might be of interest.)

ngnpope replied:
I'll try to have a look next week 🙂

FarhanAliRaza commented on django/template/backends/django.py:
```diff
@@ -75,10 +75,26 @@ def from_string(self, template_code):
         return Template(self.engine.from_string(template_code), self)
 
     def get_template(self, template_name):
+        template_name, _, partial_name = template_name.partition("#")
```
Within GitHub action, it raises a permission error because  I  think it tries to read from the home directory. 
So I converted to check for the name instead of handling the  `IsADirectoryError` error.

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,171 @@ def do_with(parser, token):
+            partials_content = self.parent_dict[self.lookup_key]
+        except KeyError:
+            raise TemplateSyntaxError(
+                f"No partials are defined. You are trying to access '{key}' partial"
```
```suggestion
                f"Partial '{key}' could not be found because no partials are defined."
```

FarhanAliRaza replied:
There is a slight difference in these errors. The Key error you mentioned occurs when no partials are defined at all. The below KeyError occurs when some partials are defined, but the specific partial user is asking for is not defined. So I added that error to show the user that no partials are defined at all. 
I can change it if we want the same error. I understand that both error means the same. This just gives an extra hint that you have not defined any partials.

FarhanAliRaza replied:
Incorporated the change having the same error message makes testing easier. If we want, we can refactor if we want different error messages.

nessita commented on django/template/defaulttags.py:
```diff
@@ -1564,3 +1567,171 @@ def do_with(parser, token):
+            return partials_content[key]
+        except KeyError:
+            raise TemplateSyntaxError(
+                f"You are trying to access an undefined partial '{key}'"
```
```suggestion
                f"Partial '{key}' is not defined in the current template."
```

carltongibson commented:
@nessita I don't know why GitHub is not showing it on the main Conversations tab, but I replied to your https://github.com/django/django/pull/19643#discussion_r2219875687, and it does show in the Files view.

FarhanAliRaza commented:
Extra review points are fixed. Now, I'm just waiting for the code-moving bit to be finalized.

nessita reviewed (commented):
Astonishing progress, thank you @FarhanAliRaza and @carltongibson!

FarhanAliRaza commented:
Added syntax tests. Refactored tests to match other syntax tests like `include` tests. 
Moved `TemplateProxy` to `base.py` and renamed it to `PartialTemplate`.
Moved `SubDictionaryWrapper` to `utils.py`

nessita commented:
> Added syntax tests. Refactored tests to match other syntax tests like `include` tests. Moved `TemplateProxy` to `base.py` and renamed it to `PartialTemplate`. Moved `SubDictionaryWrapper` to `utils.py`

Thank you @FarhanAliRaza for the updates, I'll be restarting work/review on this PR tomorrow morning.

nessita commented:
@ngnpope Hello! Is this PR something you wanted/could take a look? Non worries if not, but let me know when you can.

carltongibson commented:
> I didn't test this branch, ... 

OK, I tested I can't reproduce this @nessita. With the `"alsonotthere"` version, I get the expected output in the browser. 

<img width="1283" height="1049" alt="Screenshot 2025-07-30 at 21 43 25" src="https://github.com/user-attachments/assets/a078b82b-c8e3-4fa7-a6f6-700d00ba00d0" />

nessita commented:
> > I didn't test this branch, ...
> 
> OK, I tested. I can't reproduce this @nessita. With the `"alsonotthere"` version, I get the expected output in the browser.

Thank you! I'll debug more tomorrow, will start from scratch with your project.

FarhanAliRaza commented:
> @FarhanAliRaza See if you can reproduce.

@carltongibson  I can not reproduce it. Working as expected.

> TemplateNotFound for a missing partial name: #not-a-partial say. (Expected)
> <p>alsonotthere</p> output for <p>{{ nonexistent|default:"alsonotthere" }}</p> (Expected)
> VariableDoesNotExist for <p>{{ nonexistent|default:alsonotthere }}</p> (I guess we're saying Expected right? — Didn't  >think about it fully, but going by what Natalia said.)

The same applies to me for this current branch. 

@nessita Working on the test case.

nessita commented:
> @nessita I have added tests for `get_exception_info` by using the `template_debug` that I believe is the output of the `get_exception_info`
> 
> ```python
> with self.assertRaises(VariableDoesNotExist) as cm:
>             template.render(context)
> 
> exc_info = cm.exception.template_debug
> ```

Thank you! I can't look right away because I'm deep in other review, but could you please double check that if the implementation for `PartialTemplate.get_exception_info` would get changed to `return {}`, the proper test fail with meaningful information?

FarhanAliRaza commented:
@nessita 

Yes, I made sure it gave KeyError 
Should I add asserts to provide more descriptive errors?

```
ERROR: test_partial_runtime_error_exception_info (template_tests.syntax_tests.test_partials.PartialTagTests.test_partial_runtime_error_exception_info)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/farhan/code/django/tests/template_tests/syntax_tests/test_partials.py", line 448, in test_partial_runtime_error_exception_info
    self.assertIn("badsimpletag", exc_debug["during"])
                                  ~~~~~~~~~^^^^^^^^^^
KeyError: 'during'

======================================================================
ERROR: test_partial_runtime_exception_has_debug_info (template_tests.syntax_tests.test_partials.PartialTagTests.test_partial_runtime_exception_has_debug_info)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/farhan/code/django/tests/template_tests/syntax_tests/test_partials.py", line 303, in test_partial_runtime_exception_has_debug_info
    self.assertEqual(exc_info["during"], "{{ nonexistent|default:alsonotthere }}")
                     ~~~~~~~~^^^^^^^^^^
KeyError: 'during'

======================================================================
FAIL: test_partial_template_get_exception_info_delegation (template_tests.syntax_tests.test_partials.PartialTagTests.test_partial_template_get_exception_info_delegation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/farhan/code/django/tests/template_tests/syntax_tests/test_partials.py", line 335, in test_partial_template_get_exception_info_delegation
    self.assertIn("message", exc_info)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
AssertionError: 'message' not found in {}

----------------------------------------------------------------------
Ran 25 tests in 0.016s
```

ngnpope commented on django/template/defaulttags.py:
```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    tokens = token.split_contents()
+
+    # Check the number of tokens before trying to assign them via indexes.
+    if (tokens_len := len(tokens)) not in (2, 3):
+        name = token.contents.split()[0]
+        raise TemplateSyntaxError(f"{name} tag requires 2-3 arguments")
+
+    partial_name = tokens[1]
+    if tokens_len > 2:
+        inline = tokens[2]
+        if inline != "inline":
+            raise TemplateSyntaxError(
+                "The 'inline' argument does not have any parameters; "
+                "either use 'inline' or remove it completely."
+            )
+    else:
+        inline = False
```
Can this be less convoluted by using a match statement?

```suggestion
    match token.split_contents():
        case "partialdef", partial_name, "inline":
            inline = True
        case "partialdef", partial_name, _:
            raise TemplateSyntaxError(
                "The 'inline' argument does not have any parameters; "
                "either use 'inline' or remove it completely."
            )
        case "partialdef", partial_name:
            inline = False
        case _:
            raise TemplateSyntaxError("partialdef tag requires 2-3 arguments")
```

I find this much easier to read.

I realise that this hard-codes the "partialdef" name, but that's something that we can fix if need be - I have ideas of how to tweak the above...

carltongibson replied:
> I realise that this hard-codes the "partialdef" name

I don't think that's a problem. The chances of a tag rename are pretty low. 

(template-partials originally had `startpartial`, and `endpartial`, which didn't match Django's `f"end{tokens[0]}"` naming convention, so we updated to `partialdef` and `endpartialdef` but had to parameterise for backwards compatibility with the original names. We don't need that still.)

ngnpope commented on django/template/defaulttags.py:
```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
+        inline = False
+
+    # Parse the content until the end tag.
+    acceptable_endpartials = ("endpartialdef", f"endpartialdef {partial_name}")
```
...but it does highlight that if the tag name is supposed to be dynamic, then this hard-coding of `"endpartialdef"` is wrong and it should be `f"end{tokens[0]}"`.

Further, if it's acceptable to pass in the `partial_name` to the end tag, then we also need to fix `partial_end_tag_re` and perhaps also do some checking in `PartialTemplate.find_partial_source()`? Although the latter would likely be complex, e.g. what if someone has tried to nest `partialdef`s? Is that even supported/valid? I'm guessing from a quick scan of the lines below this it isn't supported. Do we have any tests for that?

carltongibson replied:
> Although the latter would likely be complex, e.g. what if someone has tried to nest partialdefs? Is that even supported/valid?

Nested partials **are** supported. This kind of thing is common: 

```html+django
{% extends "base.html" %}

{% block content %}

This is my main page. 

{% partialdef outer inline %}
    It hosts a couple of partials. An outer one... 
    
    (Perhaps a whole list)
    
    {% partialdef inner inine %}
        ... and an inner one. 
        
        (Perhaps just a single list item)
        
    {% endpartialdef inner %}
{% endpartialdef outer %}

{% endblock content %}
```

nessita replied:
@ngnpope Could you please clarify what do you mean with "it does highlight that if the tag name is supposed to be dynamic, then this hard-coding of "endpartialdef" is wrong and it should be f"end{tokens[0]}"."?

ngnpope commented on django/template/defaulttags.py:
```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
+
+        {% partial partial_name %}
+    """
+    bits = token.split_contents()
+    if len(bits) != 2:
+        raise TemplateSyntaxError(f"'{bits[0]}' tag requires a single argument")
+    tag_name, partial_name = bits
+
+    extra_data = getattr(parser, "extra_data")
+    partial_mapping = SubDictionaryWrapper(extra_data, "template-partials")
+
+    return RenderPartialNode(partial_name, partial_mapping=partial_mapping)
```
Again, could simplify with a match statement:

```suggestion
    match token.split_contents():
        case "partial", partial_name:
            extra_data = getattr(parser, "extra_data")
            partial_mapping = SubDictionaryWrapper(extra_data, "template-partials")
            return RenderPartialNode(partial_name, partial_mapping=partial_mapping)
        case _:
            raise TemplateSyntaxError("'partial' tag requires a single argument")
```

ngnpope commented on docs/ref/templates/builtins.txt:
```diff
@@ -957,6 +957,76 @@ output (as a string) inside a variable. This is useful if you want to use
+    {% endpartialdef %}
+
+The ``{% partialdef %}`` tag takes a required ``partial_name`` argument, which
+is a identifier for the partial:
```
```suggestion
is an identifier for the partial:
```

ngnpope commented on docs/ref/templates/language.txt:
```diff
@@ -475,6 +475,150 @@ it also defines the content that fills the hole in the *parent*. If there were
+
+.. code-block:: html+django
+
+    {% partialdef user-info %}
+        <div class="user">
+            <span>{{ user.username }}</span>
+            <span>{{ user.email }}</span>
+        </div>
+    {% endpartialdef %}
+
+    {% block content %}
+        <h2>Authors</h2>
+        {% for user in authors %}
+            {% partial user-info %}
+        {% endfor %}
+
+        <h2>Editors</h2>
+        {% for user in editors %}
+            {% partial user-info %}
+        {% endfor %}
+    {% endblock %}
```
This is a nice example! 🙌🏻

ngnpope commented on docs/ref/templates/api.txt:
```diff
@@ -158,6 +158,29 @@ overridden by what's passed by
     Loads a template with the given name, compiles it and returns a
     :class:`Template` object.
 
+    **Partial loading:**
+
+    To load a specific partial from a template:
+
+    .. code-block:: python
+
+        # Load an entire template.
+        template = engine.get_template("template.html")
+
+        # Load a specific fragment from a template.
+        partial = engine.get_template("template.html#partial_name")
+
+    When loading partials, this method returns a template object that behaves 
+    like a regular :class:`Template` but contains only the partial's content.
+
+    See :ref:`template-partials` for more information about defining and using
+    template partials.
+
+    .. versionchanged:: 6.0
+
+        Added support for loading template partials using the
+        ``template_name#partial_name`` syntax.
+
```
Is this in the right place? This is the generic Engine API documentation and this feature only works for Django templates and not Jinja2 templates, right?

ngnpope commented on docs/ref/templates/language.txt:
```diff
@@ -475,6 +475,150 @@ it also defines the content that fills the hole in the *parent*. If there were
+    {% endwith %}
+
+
+.. admonition:: Partial name uniqueness
+
+    You cannot define multiple partials with the same name in the same template.
+    If you do, the last partial definition will overwrite previous ones and be
+    the one that gets rendered.
```
Is there a viable reason to do allow this? Should we error instead so that issues can be picked up during development? It feels like this would be something that people wouldn't want to silently happen.

FarhanAliRaza replied:
Now it raises a TemplateSyntaxError during parsing. 
```python
template_source = """
        {% partialdef duplicate %}DUPLICATE-CONTENT{% endpartialdef %}
        {% partialdef duplicate %}DUPLICATE-CONTENT-2{% endpartialdef %}
        """
with self.assertRaisesMessage(
    TemplateSyntaxError,
    "Partial 'duplicate' is already defined in the 'template.html' template.",
):
    Template(template_source, origin=Origin(name="template.html"))
```

FarhanAliRaza replied:
removed from the docs also.

ngnpope commented on django/template/defaulttags.py:
```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
+        raise TemplateSyntaxError(f"'{bits[0]}' tag requires a single argument")
+    tag_name, partial_name = bits
+
+    extra_data = getattr(parser, "extra_data")
```
I don't think this `getattr()` call is necessary as `extra_data` is always available.

```suggestion
    extra_data = parser.extra_data
```

As there is no default specified, the `getattr()` would still raise an `AttributeError` anyway.

ngnpope commented on django/template/utils.py:
```diff
@@ -94,6 +96,33 @@ def all(self):
+        self.lookup_key = lookup_key
+
+    def __getitem__(self, key):
+        try:
+            partials_content = self.parent_dict[self.lookup_key]
+        except KeyError:
+            raise TemplateSyntaxError(
+                f"Partial '{key}' is not defined in the current template."
+            )
+
+        try:
+            return partials_content[key]
+        except KeyError:
+            raise TemplateSyntaxError(
+                f"Partial '{key}' is not defined in the current template."
+            )
```
The two exception blocks are the same, so I think this could be flattened, e.g.

```suggestion
        try:
            return self.parent_dict[self.lookup_key][key]
        except KeyError:
            raise TemplateSyntaxError(
                f"Partial '{key}' is not defined in the current template."
            )
```

But I think that we should remove the exception handling here and move it to `RenderPartialNode.render()`:

```python
    def render(self, context):
        try:
            return self.partial_mapping[self.partial_name].render(context)
        except KeyError:
            raise TemplateSyntaxError(
                f"Partial '{self.partial_name}' is not defined in the current template."
            )
```

And then we should move `SubDictionaryWrapper` to `django.utils.datastructures` with a more simple non-template focused implementation:

```python
class SubDictionaryWrapper:
    """Wrap a parent dictionary, allowing deferred access to a sub-dict by key."""

    def __init__(self, parent_dict, deferred_key):
        self.parent_dict = parent_dict
        self.deferred_key = deferred_key

    def __getitem__(self, key):
        return self.parent_dict[self.deferred_key][key]
```

(I renamed `lookup_key` to `deferred_key` which feels like it better reflects what it is.)

I tried to think of another way around this, but the only thing that came to mind was passing `extra-data` down into `RenderPartialNode` which should then have a reference to the object that is populated with `template-partials` later, but that feels wrong.

FarhanAliRaza commented on django/template/defaulttags.py:
```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
+
+    partial_name = tokens[1]
+    if tokens_len > 2:
+        inline = tokens[2]
```
changed to this 
```
        inline = tokens[2] == "inline"
        if not inline:
            raise TemplateSyntaxError(
                "The 'inline' argument does not have any parameters; "
                "either use 'inline' or remove it completely."
            )
    else:
        inline = False
```

FarhanAliRaza commented:
Implemented the suggestion from @ngnpope .
Handled nested partials handling by updating `find_partial_source` to use a single regex and handle search in nested case.
Refactored  `SubDictionaryWrapper` to move error handling out of it to  `RenderPartialNode`
and moved it to `django/utils/datastructures.py`
And other refactors.

nessita reviewed (commented):
Thank you @FarhanAliRaza! This is shaping nicely and with a delight precision. :star2: 

Today I focused mostly on the new syntax tests. Overall I would like to see more meaningful names used for the tests and the templates, and also more of `subTest` when semantically adequate. I added some suggestions for both of these points.

Then, a minor nitpick: I think that the partial name used for testing could be `partial-name` or `testing-name` to signal clearly to readers that this is a name identifier.

Thinking about performance, we may need one or two tests with big chunky templates (using html test files) making heavy use of the new tags to see if the regexes show any obvious catastrophic side-effect.

nessita commented:
> @FarhanAliRaza @nessita on [the new-features repo there's a suggestion](https://github.com/django/new-features/issues/66#issuecomment-3161031954) to move `_render()` and `render()` into a common base mixin, that we might want to pick up here?

I have read the new feature request and caught up with the comments in the django-components issue. My gut feeling is that this is a good idea and it'd solve somehow nicely the itchiness that the duplicated `render`/`_render` methods is causing me.

There is the outstanding question I just posted in my latest review about whether the `PartialTemplate._render` method is used at all (since making it return `""` does not have any test failing).

carltongibson commented:
> Thinking about performance, we may need one or two tests with big chunky templates (using html test files) making heavy use of the new tags to see if the regexes show any obvious catastrophic side-effect.

What do you have in mind here @nessita? (Im struggling to imagine the kind of case you want to look at.)

The regex is just used for the debug view right, fetching the source. Is there a DoS vector there? 🤔

FarhanAliRaza commented:
@nessita  Now tests will fail if we change `_render` method because even after instrumentation, it calls the original method.

FarhanAliRaza commented on django/test/utils.py:
```diff
@@ -111,6 +111,9 @@ def instrumented_test_render(self, context):
     intercepted by the test Client.
     """
     template_rendered.send(sender=self, template=self, context=context)
+    if isinstance(self, PartialTemplate):
+        return _TestState.saved_data.partial_template_render(self, context)
+
```
So add a test `test_private_render_is_actually_called`. 
or
We do the same solution with `Template` class. 
```return _TestState.saved_data.template_render(self, context)```
As this makes more sense because this way always the original function is being called, we are getting the signal that we want from instrumenting. It will make it easier to test any future changes if a contributor makes any changes, and they will see the changes during testing. 
If this causes no other unexpected problems ...

carltongibson replied:
Option 3 is to do nothing, but a regression test is likely worthwhile.

I **don't like** the suggested change here. If we want to go this kind of route then we want to adjust `instrumented_test_render` to decorate the original method, and apply that, rather than using a pair of `isinstance` checks.

Given how long `instrumented_test_render` has worked how it has, I'd suggest that's a separate cleanup not part of this PR.  (Would it not need a ticket?)

FarhanAliRaza replied:
If we decide to go to the `Renderer` Mixin route to refactor `_render` and `render` method to it. We will not even need that isinstance check (I think).

ngnpope commented on django/template/base.py:
```diff
@@ -88,6 +88,12 @@
 # than instantiating SimpleLazyObject with _lazy_re_compile().
 tag_re = re.compile(r"({%.*?%}|{{.*?}}|{#.*?#})")
 
+combined_partial_pattern = (
+    r"(\{% \s* partialdef \s+ (?P<name>[\w-]+) (?:\s+ [^%]*)? \s* %\})"
+    r"|(\{% \s* endpartialdef (?:\s+ [^%]*)? \s* %\})"
+)
+combined_partial_re = re.compile(combined_partial_pattern, re.VERBOSE)
```
I think we can both simplify the regex and make it more explicit:

```suggestion
combined_partial_re = re.compile(
    r"{%\s*partialdef\s+(?P<name>[\w-]+)(?:\s+inline)?\s*%}"
    r"|{%\s*endpartialdef(?:\s+[\w-]+)?\s*%}"
)
```

ngnpope commented on django/template/base.py:
```diff
@@ -288,6 +294,59 @@ def get_exception_info(self, exception, token):
+        start_match = None
+        nesting = 0
+
+        for match in combined_partial_re.finditer(full_source):
+            if match.group(1):  # Start tag
+                tag_name = match.group("name")
+                if tag_name == partial_name and start_match is None:
+                    start_match = match
+                    nesting = 1
+                elif start_match is not None:
+                    nesting += 1
+            else:  # End tag (group 2)
+                if start_match is not None:
+                    nesting -= 1
+                    if nesting == 0:
+                        return full_source[start_match.start() : match.end()]
```
This can be a little less verbose:

```suggestion
        for match in combined_partial_re.finditer(full_source):
            if match["name"]:  # Start Tag
                if start_match is None and match["name"] == partial_name:
                    start_match = match
                if start_match is not None:
                    nesting += 1
            elif start_match is not None:
                nesting -= 1
                if nesting == 0:
                    return full_source[start_match.start() : match.end()]
```

FarhanAliRaza commented on tests/template_tests/syntax_tests/test_partials.py:
```diff
@@ -0,0 +1,537 @@
+        {
+            "partial10": (
+                "MAIN TEMPLATE START "
+                "{% include 'partial_included.html' %} "
```
It is defined at the top in `partial_templates`.

FarhanAliRaza commented on tests/template_tests/syntax_tests/test_partials.py:
```diff
@@ -0,0 +1,537 @@
+        self.assertIn("INCLUDED TEMPLATE END", output)
+        self.assertIn("MAIN TEMPLATE END", output)
+
+    # Error cases
```
Removed as it is testes already.

nessita commented:
> > Thinking about performance, we may need one or two tests with big chunky templates (using html test files) making heavy use of the new tags to see if the regexes show any obvious catastrophic side-effect.
> 
> What do you have in mind here @nessita? (Im struggling to imagine the kind of case you want to look at.)

I'll try to think of a concrete example and I'll share here.

> The regex is just used for the debug view right, fetching the source. Is there a DoS vector there? 🤔

Good point, while going deep in the review vertical I lost track of this detail. Can you help me understand which flow exactly triggers this code path?

FarhanAliRaza commented:
>The tests are the ones below and using the {% include 'partial_included.html#included-partial' %} should work, right?

I think when the `include` tag is used, it tries to load using the base `Engine` instance, which does not use `Django Template Backend` 

Some options that I can think of to fix this are to change 

`Engine.get_template` method to make it aware of partials.

Change the loading logic in the `Include` tag to somehow load using the django backend.

carltongibson commented:
@FarhanAliRaza The examples from those tests work standardly. I use that pattern **a lot**. Wondering if it's a test artifact? 🤔

The base `Engine.get_template` goes via the loader, which traverses the backends to find the correct template. This still goes via the Django backend, which is partial aware.

FarhanAliRaza commented:
I tried without @setup decorator also .
It gives the same error .
I changed it load using filesystem template files.

But in `include` the loading happens through context.template loading that somehow does not use the Django backend. I am not sure how exactly but I added print statement loading during include, it does not Django backend for that.

FarhanAliRaza commented:
I only tested the usage of partial inside the included HTML file. Not tested partial loading like this `{% include 'partial_included.html#included-partial' %}`   that is failing now. 

only tested this about `include`
```
@setup(
        {
            "partial_with_include": (
                "MAIN TEMPLATE START\n"
                "{% include 'partial_included.html' %}\n"
                "MAIN TEMPLATE END"
            )
        },
        partial_templates,
    )
    def test_partial_in_included_template(self):
        output = self.engine.render_to_string("partial_with_include")
        expected = (
            "MAIN TEMPLATE START\nINCLUDED TEMPLATE START\n\n\n"
            "Now using the partial: \n"
            "THIS IS CONTENT FROM THE INCLUDED PARTIAL\n\n\n"
            "INCLUDED TEMPLATE END\n\nMAIN TEMPLATE END"
        )
        self.assertEqual(output, expected)
```

FarhanAliRaza commented:
[test.zip](https://github.com/user-attachments/files/21728828/test.zip)
experimental project.

carltongibson commented:
OK, we've just inserted the partial lookup in the wrong place. Confusing names strike again. `django.template.backends.base.BaseEngine` is what everything subclasses. `django.template.engine.Engine` is Django specific — so not a problem adjusting it — and where the partial lookup needs to live. 

@FarhanAliRaza If you restore and django/template/backends/django.py and implement the logic there in django/template/engine.py we'll be in the right place, I think.

carltongibson commented:
> ...  unsure about template-partials

It turned out not. Well, except in my work project of course. 😅 Since the loader (where it's implemented in template-partials) is **inside** the engine it works. The issue here was we'd move the logic **outside** the engine, which meant it only worked via the _usual_ public API, but not via the reference to the engine that the include tag uses.

FarhanAliRaza commented:
Moved `get_template` partial implementation from `DjangoTemplates.get_template` to `Engine.get_template`. Added tests that @nessita  suggested. Thank you, it was a great find. 

Fixed test mocks for the new implementation.

nessita reviewed (commented):
I think we are super close to the final line! :checkered_flag: 

I have made a full-full review today. I renamed the new utility for the deferred dict access and added a full test case for it. I went thru the docs and pushed what I think is hopefully a slightly clearer wording, and I also made all the examples in `ref/templates/language.txt` consistent with the User/Author example from the "reuse" section.

I'll give a final pass tomorrow and likely merge :star2:

smithdc1 commented:
The docs mention HTMX. 🤔

I get that HTMX is popular right now but mentioning it in Django's docs seems surprising to me. Typically, the docs avoid mentioning a particular package -- to the extent that the SC have now launched the ecosystem page on the website.

carltongibson commented:
I don't think we have to pretend Django is the only software that exists. 

The docs contain [mentions of jQuery](https://docs.djangoproject.com/en/5.2/topics/forms/media/#order-of-assets), for example, and it's perfectly natural.

nessita commented:
@carltongibson @FarhanAliRaza @ngnpope Last night I dreamt about partials (no kidding), and I woke up thinking that the `extra_data` internal dict key might be better named simply `partials` (instead of `template-partials`), since the code where this key is used is already within a "template" context.

If you think this is inaccurate or a bad idea, please let me know!

carltongibson commented:
@nessita no problem, yes 👍 -- it's gotta be close if you're dreaming about it 🛌 -- thanks for all your sterling work! 🎩

FarhanAliRaza commented:
I also think your dream should be fulfilled. @nessita 😅
🎉
